File: typescript_lexer/src/Lexer.test.ts
import { Lexer } from './Lexer';
import { TokenType } from './Token';
import { LexerError } from './LexerError';

describe('Lexer', () => {
  describe('Basic Tokenization', () => {
    test('should tokenize numbers', () => {
      const lexer = new Lexer('42 3.14');
      const tokens = lexer.tokenize();

      expect(tokens).toHaveLength(3); // 2 numbers + EOF
      expect(tokens[0].type).toBe(TokenType.NUMBER);
      expect(tokens[0].value).toBe('42');
      expect(tokens[1].type).toBe(TokenType.NUMBER);
      expect(tokens[1].value).toBe('3.14');
      expect(tokens[2].type).toBe(TokenType.EOF);
    });

    test('should tokenize strings with double quotes', () => {
      const lexer = new Lexer('"hello world"');
      const tokens = lexer.tokenize();

      expect(tokens).toHaveLength(2); // 1 string + EOF
      expect(tokens[0].type).toBe(TokenType.STRING);
      expect(tokens[0].value).toBe('hello world');
    });

    test('should tokenize strings with single quotes', () => {
      const lexer = new Lexer("'hello world'");
      const tokens = lexer.tokenize();

      expect(tokens).toHaveLength(2);
      expect(tokens[0].type).toBe(TokenType.STRING);
      expect(tokens[0].value).toBe('hello world');
    });

    test('should handle escape sequences in strings', () => {
      const lexer = new Lexer('"hello\\nworld\\ttab"');
      const tokens = lexer.tokenize();

      expect(tokens[0].value).toBe('hello\nworld\ttab');
    });

    test('should tokenize identifiers', () => {
      const lexer = new Lexer('foo bar_baz _underscore');
      const tokens = lexer.tokenize();

      expect(tokens).toHaveLength(4); // 3 identifiers + EOF
      expect(tokens[0].type).toBe(TokenType.IDENTIFIER);
      expect(tokens[0].value).toBe('foo');
      expect(tokens[1].type).toBe(TokenType.IDENTIFIER);
      expect(tokens[1].value).toBe('bar_baz');
      expect(tokens[2].type).toBe(TokenType.IDENTIFIER);
      expect(tokens[2].value).toBe('_underscore');
    });
  });

  describe('Keywords', () => {
    test('should tokenize keywords', () => {
      const lexer = new Lexer('if else while for function return');
      const tokens = lexer.tokenize();

      expect(tokens[0].type).toBe(TokenType.IF);
      expect(tokens[1].type).toBe(TokenType.ELSE);
      expect(tokens[2].type).toBe(TokenType.WHILE);
      expect(tokens[3].type).toBe(TokenType.FOR);
      expect(tokens[4].type).toBe(TokenType.FUNCTION);
      expect(tokens[5].type).toBe(TokenType.RETURN);
    });

    test('should tokenize boolean and null keywords', () => {
      const lexer = new Lexer('true false null');
      const tokens = lexer.tokenize();

      expect(tokens[0].type).toBe(TokenType.TRUE);
      expect(tokens[1].type).toBe(TokenType.FALSE);
      expect(tokens[2].type).toBe(TokenType.NULL);
    });

    test('should tokenize let and const', () => {
      const lexer = new Lexer('let const');
      const tokens = lexer.tokenize();

      expect(tokens[0].type).toBe(TokenType.LET);
      expect(tokens[1].type).toBe(TokenType.CONST);
    });
  });

  describe('Operators', () => {
    test('should tokenize arithmetic operators', () => {
      const lexer = new Lexer('+ - * / %');
      const tokens = lexer.tokenize();

      expect(tokens[0].type).toBe(TokenType.PLUS);
      expect(tokens[1].type).toBe(TokenType.MINUS);
      expect(tokens[2].type).toBe(TokenType.MULTIPLY);
      expect(tokens[3].type).toBe(TokenType.DIVIDE);
      expect(tokens[4].type).toBe(TokenType.MODULO);
    });

    test('should tokenize comparison operators', () => {
      const lexer = new Lexer('< > <= >= == !=');
      const tokens = lexer.tokenize();

      expect(tokens[0].type).toBe(TokenType.LESS_THAN);
      expect(tokens[1].type).toBe(TokenType.GREATER_THAN);
      expect(tokens[2].type).toBe(TokenType.LESS_EQUAL);
      expect(tokens[3].type).toBe(TokenType.GREATER_EQUAL);
      expect(tokens[4].type).toBe(TokenType.EQUALS);
      expect(tokens[5].type).toBe(TokenType.NOT_EQUALS);
    });

    test('should tokenize logical operators', () => {
      const lexer = new Lexer('&& || !');
      const tokens = lexer.tokenize();

      expect(tokens[0].type).toBe(TokenType.AND);
      expect(tokens[1].type).toBe(TokenType.OR);
      expect(tokens[2].type).toBe(TokenType.NOT);
    });

    test('should tokenize assignment operator', () => {
      const lexer = new Lexer('=');
      const tokens = lexer.tokenize();

      expect(tokens[0].type).toBe(TokenType.ASSIGN);
    });
  });

  describe('Delimiters', () => {
    test('should tokenize parentheses and braces', () => {
      const lexer = new Lexer('( ) { } [ ]');
      const tokens = lexer.tokenize();

      expect(tokens[0].type).toBe(TokenType.LPAREN);
      expect(tokens[1].type).toBe(TokenType.RPAREN);
      expect(tokens[2].type).toBe(TokenType.LBRACE);
      expect(tokens[3].type).toBe(TokenType.RBRACE);
      expect(tokens[4].type).toBe(TokenType.LBRACKET);
      expect(tokens[5].type).toBe(TokenType.RBRACKET);
    });

    test('should tokenize punctuation', () => {
      const lexer = new Lexer('; , .');
      const tokens = lexer.tokenize();

      expect(tokens[0].type).toBe(TokenType.SEMICOLON);
      expect(tokens[1].type).toBe(TokenType.COMMA);
      expect(tokens[2].type).toBe(TokenType.DOT);
    });
  });

  describe('Comments', () => {
    test('should skip line comments', () => {
      const lexer = new Lexer('let x = 10; // this is a comment\nlet y = 20;');
      const tokens = lexer.tokenize();

      expect(tokens).toHaveLength(11); // let x = 10 ; let y = 20 ; EOF
      expect(tokens[0].type).toBe(TokenType.LET);
      expect(tokens[0].value).toBe('let');
      expect(tokens[5].type).toBe(TokenType.LET);
      expect(tokens[5].value).toBe('let');
    });

    test('should skip block comments', () => {
      const lexer = new Lexer('let x /* comment */ = 10;');
      const tokens = lexer.tokenize();

      expect(tokens).toHaveLength(6); // let x = 10 ; EOF
      expect(tokens[0].type).toBe(TokenType.LET);
      expect(tokens[1].type).toBe(TokenType.IDENTIFIER);
      expect(tokens[2].type).toBe(TokenType.ASSIGN);
    });

    test('should handle multi-line block comments', () => {
      const lexer = new Lexer(`let x = 10;
/* This is a
   multi-line
   comment */
let y = 20;`);
      const tokens = lexer.tokenize();

      expect(tokens[0].type).toBe(TokenType.LET);
      expect(tokens[5].type).toBe(TokenType.LET);
    });
  });

  describe('Complex Expressions', () => {
    test('should tokenize a simple function', () => {
      const source = `function add(a, b) {
  return a + b;
}`;
      const lexer = new Lexer(source);
      const tokens = lexer.tokenize();

      expect(tokens[0].type).toBe(TokenType.FUNCTION);
      expect(tokens[1].type).toBe(TokenType.IDENTIFIER);
      expect(tokens[1].value).toBe('add');
      expect(tokens[2].type).toBe(TokenType.LPAREN);
      expect(tokens[3].type).toBe(TokenType.IDENTIFIER);
      expect(tokens[3].value).toBe('a');
      expect(tokens[4].type).toBe(TokenType.COMMA);
    });

    test('should tokenize array initialization', () => {
      const lexer = new Lexer('const arr = [1, 2, 3];');
      const tokens = lexer.tokenize();

      expect(tokens[0].type).toBe(TokenType.CONST);
      expect(tokens[1].type).toBe(TokenType.IDENTIFIER);
      expect(tokens[2].type).toBe(TokenType.ASSIGN);
      expect(tokens[3].type).toBe(TokenType.LBRACKET);
      expect(tokens[4].type).toBe(TokenType.NUMBER);
      expect(tokens[5].type).toBe(TokenType.COMMA);
      // const arr = [ 1 , 2 , 3 ] ;
      // 0     1   2 3 4 5 6 7 8 9 10 11
      expect(tokens[9].type).toBe(TokenType.RBRACKET);
    });

    test('should tokenize conditional statement', () => {
      const source = `if (x > 10) {
  return true;
}`;
      const lexer = new Lexer(source);
      const tokens = lexer.tokenize();

      expect(tokens[0].type).toBe(TokenType.IF);
      expect(tokens[1].type).toBe(TokenType.LPAREN);
      expect(tokens[2].type).toBe(TokenType.IDENTIFIER);
      expect(tokens[3].type).toBe(TokenType.GREATER_THAN);
      expect(tokens[4].type).toBe(TokenType.NUMBER);
    });
  });

  describe('Error Handling', () => {
    test('should throw error for unterminated string', () => {
      const lexer1 = new Lexer('"unterminated');
      const lexer2 = new Lexer('"unterminated');

      expect(() => lexer1.tokenize()).toThrow(LexerError);
      expect(() => lexer2.tokenize()).toThrow('Unterminated string literal');
    });

    test('should throw error for invalid character', () => {
      const lexer1 = new Lexer('let x = 10 @ 5;');
      const lexer2 = new Lexer('let x = 10 @ 5;');

      expect(() => lexer1.tokenize()).toThrow(LexerError);
      expect(() => lexer2.tokenize()).toThrow('Unexpected character');
    });

    test('should throw error for single ampersand', () => {
      const lexer = new Lexer('let x = a & b;');

      expect(() => lexer.tokenize()).toThrow(LexerError);
    });

    test('should throw error for single pipe', () => {
      const lexer = new Lexer('let x = a | b;');

      expect(() => lexer.tokenize()).toThrow(LexerError);
    });

    test('should throw error for unterminated block comment', () => {
      const lexer1 = new Lexer('/* unterminated comment');
      const lexer2 = new Lexer('/* unterminated comment');

      expect(() => lexer1.tokenize()).toThrow(LexerError);
      expect(() => lexer2.tokenize()).toThrow('Unterminated block comment');
    });

    test('should provide line and column information in errors', () => {
      const lexer = new Lexer('let x = 10;\nlet y = @;');

      try {
        lexer.tokenize();
        fail('Should have thrown an error');
      } catch (error) {
        expect(error).toBeInstanceOf(LexerError);
        if (error instanceof LexerError) {
          expect(error.line).toBe(2);
          expect(error.column).toBeGreaterThan(0);
        }
      }
    });
  });

  describe('Token Methods', () => {
    test('Token.is() should check token type', () => {
      const lexer = new Lexer('let');
      const tokens = lexer.tokenize();

      expect(tokens[0].is(TokenType.LET)).toBe(true);
      expect(tokens[0].is(TokenType.CONST)).toBe(false);
    });

    test('Token.isOneOf() should check multiple types', () => {
      const lexer = new Lexer('let const if');
      const tokens = lexer.tokenize();

      expect(tokens[0].isOneOf([TokenType.LET, TokenType.CONST])).toBe(true);
      expect(tokens[1].isOneOf([TokenType.LET, TokenType.CONST])).toBe(true);
      expect(tokens[2].isOneOf([TokenType.LET, TokenType.CONST])).toBe(false);
    });

    test('Token.toString() should return formatted string', () => {
      const lexer = new Lexer('let');
      const tokens = lexer.tokenize();

      const str = tokens[0].toString();
      expect(str).toContain('Token');
      expect(str).toContain('LET');
      expect(str).toContain('let');
    });
  });

  describe('Recursion Demonstration', () => {
    test('should handle recursive digit consumption', () => {
      const lexer = new Lexer('123456789');
      const tokens = lexer.tokenize();

      expect(tokens[0].type).toBe(TokenType.NUMBER);
      expect(tokens[0].value).toBe('123456789');
    });

    test('should handle decimal numbers with recursion', () => {
      const lexer = new Lexer('3.14159265');
      const tokens = lexer.tokenize();

      expect(tokens[0].type).toBe(TokenType.NUMBER);
      expect(tokens[0].value).toBe('3.14159265');
    });
  });
});

File: erlang_lexer/test/lexer_test.erl
-module(lexer_test).
-include_lib("eunit/include/eunit.hrl").
-include("token.hrl").

%% Helper: tokenize and return the token list (asserting success).
tokenize(Source) ->
    {ok, Tokens} = lexer:tokenize(Source),
    Tokens.

%% Helper: return tokens excluding the trailing EOF.
tokens_no_eof(Source) ->
    Tokens = tokenize(Source),
    lists:filter(fun(#token{type = T}) -> T =/= eof end, Tokens).

%% ===================================================================
%% Basic Tokenization
%% ===================================================================

number_integer_test() ->
    [#token{type = number, value = "42"} | _] = tokenize("42").

number_float_test() ->
    [#token{type = number, value = "3.14"} | _] = tokenize("3.14").

string_double_quote_test() ->
    [#token{type = string, value = "hello"} | _] = tokenize("\"hello\"").

string_single_quote_test() ->
    [#token{type = string, value = "world"} | _] = tokenize("'world'").

identifier_test() ->
    [#token{type = identifier, value = "myVar"} | _] = tokenize("myVar").

identifier_with_underscore_test() ->
    [#token{type = identifier, value = "_private"} | _] = tokenize("_private").

%% ===================================================================
%% Keywords
%% ===================================================================

keyword_if_test() ->
    [#token{type = 'if', value = "if"} | _] = tokenize("if").

keyword_else_test() ->
    [#token{type = 'else', value = "else"} | _] = tokenize("else").

keyword_while_test() ->
    [#token{type = 'while', value = "while"} | _] = tokenize("while").

keyword_for_test() ->
    [#token{type = 'for', value = "for"} | _] = tokenize("for").

keyword_function_test() ->
    [#token{type = 'function', value = "function"} | _] = tokenize("function").

keyword_return_test() ->
    [#token{type = 'return', value = "return"} | _] = tokenize("return").

keyword_let_test() ->
    [#token{type = 'let', value = "let"} | _] = tokenize("let").

keyword_const_test() ->
    [#token{type = 'const', value = "const"} | _] = tokenize("const").

keyword_true_test() ->
    [#token{type = 'true', value = "true"} | _] = tokenize("true").

keyword_false_test() ->
    [#token{type = 'false', value = "false"} | _] = tokenize("false").

keyword_null_test() ->
    [#token{type = 'null', value = "null"} | _] = tokenize("null").

%% ===================================================================
%% Operators
%% ===================================================================

operator_plus_test() ->
    [#token{type = plus, value = "+"} | _] = tokenize("+").

operator_minus_test() ->
    [#token{type = minus, value = "-"} | _] = tokenize("-").

operator_multiply_test() ->
    [#token{type = multiply, value = "*"} | _] = tokenize("*").

operator_divide_test() ->
    [#token{type = divide, value = "/"} | _] = tokenize("/").

operator_modulo_test() ->
    [#token{type = modulo, value = "%"} | _] = tokenize("%").

operator_assign_test() ->
    [#token{type = assign, value = "="} | _] = tokenize("=").

operator_equals_test() ->
    [#token{type = equals, value = "=="} | _] = tokenize("==").

operator_not_equals_test() ->
    [#token{type = not_equals, value = "!="} | _] = tokenize("!=").

operator_less_than_test() ->
    [#token{type = less_than, value = "<"} | _] = tokenize("<").

operator_greater_than_test() ->
    [#token{type = greater_than, value = ">"} | _] = tokenize(">").

operator_less_equal_test() ->
    [#token{type = less_equal, value = "<="} | _] = tokenize("<=").

operator_greater_equal_test() ->
    [#token{type = greater_equal, value = ">="} | _] = tokenize(">=").

operator_logical_and_test() ->
    [#token{type = logical_and, value = "&&"} | _] = tokenize("&&").

operator_logical_or_test() ->
    [#token{type = logical_or, value = "||"} | _] = tokenize("||").

operator_logical_not_test() ->
    [#token{type = logical_not, value = "!"} | _] = tokenize("!").

%% ===================================================================
%% Delimiters
%% ===================================================================

delimiter_lparen_test() ->
    [#token{type = lparen} | _] = tokenize("(").

delimiter_rparen_test() ->
    [#token{type = rparen} | _] = tokenize(")").

delimiter_lbrace_test() ->
    [#token{type = lbrace} | _] = tokenize("{").

delimiter_rbrace_test() ->
    [#token{type = rbrace} | _] = tokenize("}").

delimiter_lbracket_test() ->
    [#token{type = lbracket} | _] = tokenize("[").

delimiter_rbracket_test() ->
    [#token{type = rbracket} | _] = tokenize("]").

delimiter_semicolon_test() ->
    [#token{type = semicolon} | _] = tokenize(";").

delimiter_comma_test() ->
    [#token{type = comma} | _] = tokenize(",").

delimiter_dot_test() ->
    [#token{type = dot} | _] = tokenize(".").

%% ===================================================================
%% Comments
%% ===================================================================

line_comment_test() ->
    Tokens = tokens_no_eof("42 // this is a comment"),
    ?assertEqual(1, length(Tokens)),
    [#token{type = number, value = "42"}] = Tokens.

block_comment_test() ->
    Tokens = tokens_no_eof("42 /* block comment */ 7"),
    ?assertEqual(2, length(Tokens)),
    [#token{type = number, value = "42"}, #token{type = number, value = "7"}] = Tokens.

nested_block_comment_test() ->
    Tokens = tokens_no_eof("1 /* outer /* inner */ still comment */ 2"),
    ?assertEqual(2, length(Tokens)),
    [#token{type = number, value = "1"}, #token{type = number, value = "2"}] = Tokens.

%% ===================================================================
%% Complex Expressions
%% ===================================================================

simple_expression_test() ->
    Tokens = tokens_no_eof("let x = 42 + 3.14;"),
    Types = [T#token.type || T <- Tokens],
    ?assertEqual(['let', identifier, assign, number, plus, number, semicolon], Types).

function_declaration_test() ->
    Tokens = tokens_no_eof("function add(a, b) { return a + b; }"),
    Types = [T#token.type || T <- Tokens],
    ?assertEqual(['function', identifier, lparen, identifier, comma, identifier, rparen,
                  lbrace, 'return', identifier, plus, identifier, semicolon, rbrace], Types).

conditional_test() ->
    Tokens = tokens_no_eof("if (x > 10) { return true; }"),
    Types = [T#token.type || T <- Tokens],
    ?assertEqual(['if', lparen, identifier, greater_than, number, rparen,
                  lbrace, 'return', 'true', semicolon, rbrace], Types).

%% ===================================================================
%% Position Tracking
%% ===================================================================

position_tracking_test() ->
    [T1, T2, T3 | _] = tokenize("let x"),
    ?assertEqual({1, 1}, {T1#token.line, T1#token.column}),
    ?assertEqual({1, 5}, {T2#token.line, T2#token.column}),
    %% EOF
    ?assertEqual({1, 6}, {T3#token.line, T3#token.column}).

multiline_position_test() ->
    Tokens = tokens_no_eof("let\nx"),
    [T1, T2] = Tokens,
    ?assertEqual({1, 1}, {T1#token.line, T1#token.column}),
    ?assertEqual({2, 1}, {T2#token.line, T2#token.column}).

%% ===================================================================
%% Error Cases
%% ===================================================================

unterminated_string_test() ->
    ?assertMatch({error, _}, lexer:tokenize("\"hello")).

invalid_single_ampersand_test() ->
    ?assertMatch({error, _}, lexer:tokenize("&")).

invalid_single_pipe_test() ->
    ?assertMatch({error, _}, lexer:tokenize("|")).

unexpected_character_test() ->
    ?assertMatch({error, _}, lexer:tokenize("@")).

unterminated_block_comment_test() ->
    ?assertMatch({error, _}, lexer:tokenize("/* unterminated")).

%% ===================================================================
%% String Escape Sequences
%% ===================================================================

string_escape_newline_test() ->
    [#token{type = string, value = Value} | _] = tokenize("\"hello\\nworld\""),
    ?assertEqual("hello\nworld", Value).

string_escape_tab_test() ->
    [#token{type = string, value = Value} | _] = tokenize("\"hello\\tworld\""),
    ?assertEqual("hello\tworld", Value).

string_escape_backslash_test() ->
    [#token{type = string, value = Value} | _] = tokenize("\"hello\\\\world\""),
    ?assertEqual("hello\\world", Value).

%% ===================================================================
%% Token Module Functions
%% ===================================================================

token_to_string_test() ->
    T = token:new(number, "42", 1, 1),
    Str = lists:flatten(token:to_string(T)),
    ?assert(is_list(Str)),
    ?assertNotEqual(0, length(Str)).

token_is_type_test() ->
    T = token:new(number, "42", 1, 1),
    ?assert(token:is_type(T, number)),
    ?assertNot(token:is_type(T, string)).

token_is_one_of_test() ->
    T = token:new(number, "42", 1, 1),
    ?assert(token:is_one_of(T, [number, string, identifier])),
    ?assertNot(token:is_one_of(T, [string, identifier])).

%% ===================================================================
%% EOF Token
%% ===================================================================

eof_token_test() ->
    Tokens = tokenize(""),
    ?assertEqual(1, length(Tokens)),
    [#token{type = eof}] = Tokens.

eof_at_end_test() ->
    Tokens = tokenize("42"),
    Last = lists:last(Tokens),
    ?assertEqual(eof, Last#token.type).

%% ===================================================================
%% Lexer Error Module
%% ===================================================================

lexer_error_format_test() ->
    Err = lexer_error:new("Test error", 5, 10),
    Msg = lexer_error:format_message(Err),
    ?assert(is_list(Msg)),
    ?assertNotEqual(nomatch, string:find(Msg, "Test error")),
    ?assertNotEqual(nomatch, string:find(Msg, "5")),
    ?assertNotEqual(nomatch, string:find(Msg, "10")).

lexer_error_format_with_char_test() ->
    Err = lexer_error:new("Bad char", 1, 1, "@"),
    Msg = lexer_error:format_message(Err),
    ?assertNotEqual(nomatch, string:find(Msg, "@")).
